daffodil-runtime2/src/main/scala/org/apache/daffodil/runtime2/generators/CodeGeneratorState.scala

```diff
@@ -404,76 +542,117 @@ class CodeGeneratorState(private val root: ElementBase) {
-      val unparseStatement = s"    case $offset:"
-
-      structs.top.initChoiceStatements += initChoiceStatement
-      structs.top.parserStatements += parseStatement
-      structs.top.unparserStatements += unparseStatement
+      structs.top.initChoiceStatements += s"        instance->_choice = $position;"
+      structs.top.parserStatements += s"    case $position:"
+      structs.top.unparserStatements += s"    case $position:"
     }
+    val indent1 = if (hasChoice) "    " else ""
 
-    def addStatements(deref: String): Unit = {
-      val initChoiceStatement = s"        ${C}_initERD(&instance->$e$deref);"
-      val initERDStatement = s"    ${C}_initERD(&instance->$e$deref);"
-      val initSelfStatement = s"    ${C}_initSelf(&instance->$e$deref);"
-      val moreIndent = if (hasChoice) "    " else ""
-      val parseStatement =
-        s"""$moreIndent    ${C}_parseSelf(&instance->$e$deref, pstate);
-           |$moreIndent    if (pstate->error) return;""".stripMargin
-      val unparseStatement =
-        s"""$moreIndent    ${C}_unparseSelf(&instance->$e$deref, ustate);
-           |$moreIndent    if (ustate->error) return;""".stripMargin
+    // Add an array if the child occurs multiple times
+    val hasArray = arrayMaxOccurs(child) > 0
+    val deref = if (hasArray) "[i]" else ""
+    val indent2 = if (hasArray) "    " else ""
+    if (hasArray) pushArray(child)
 
+    // Generate statements to initialize, parse, and unparse the child
+    Runtime2CodeGenerator.generateCode(child.enclosedElement, this)
+    if (!child.isSimpleType) {
       if (hasChoice)
-        structs.top.initChoiceStatements += initChoiceStatement
+        structs.top.initChoiceStatements += s"$indent2        ${C}_initERD(&instance->$e$deref, (InfosetBase *)instance);"
       else
-        structs.top.initERDStatements += initERDStatement
-      structs.top.initSelfStatements += initSelfStatement
-      structs.top.parserStatements += parseStatement
-      structs.top.unparserStatements += unparseStatement
+        structs.top.initERDStatements += s"$indent2    ${C}_initERD(&instance->$e$deref, (InfosetBase *)instance);"
+      structs.top.parserStatements +=
+        s"""$indent1$indent2    ${C}_parseSelf(&instance->$e$deref, pstate);
+           |$indent1$indent2    if (pstate->error) return;""".stripMargin
+      structs.top.unparserStatements +=
+        s"""$indent1$indent2    ${C}_unparseSelf(&instance->$e$deref, ustate);
+           |$indent1$indent2    if (ustate->error) return;""".stripMargin
     }
-    if (arraySize > 0)
-      for (i <- 0 until arraySize)
-        addStatements(s"[$i]")
-    else
-      addStatements("")
 
-    if (hasChoice) {
-      val initChoiceStatement = s"        break;"
-      val parseStatement = s"        break;"
-      val unparseStatement = s"        break;"
+    // Generate statements to align the next element
+    val mustAlign = child.endingAlignmentApprox.nBits > child.priorAlignmentApprox.nBits
```

Comment:
> there still was a need for extra alignment between two elements aligned to 8 bits
because the schema writer had used an empty sequence aligned to 32
bits following a hexBinary element to artificially pad the hexBinary
element's length to a multiple of 32 bits, and furthermore, the schema
compiler had optimized away that empty sequence so the only place that
extra 32-bit alignment could be found was in the hexBinary element's
endingAlignmentApprox.nBits (32) which was bigger than its
priorAlignmentApprox.nBits (8) and its alignmentValueInBits (8).

I'm not sure why DFDL would have optimized out the sequence if there's required alignment. That sounds like it might be a bug in the compiler if that's the case.

It feels like runtime2 should be inserting framing code for sequences that are not known to be aligned, in addition to elements. Would it work to add the same `isKnownToBeAligned` and `parse_align` logic in `orderedSequeneGenrateConfig` or wherever runtime2 walks into sequences?

Reply:
I'd like to suggest that Daffodil's schema compiler should provide an available option which is to require all alignment to be static except for alignment after something that is truly variable length. Runtime 2 would then always require this to be enabled. But I think this would be popular for any schema author. 

For example: 

this would not be allowed:
```
<sequence dfdl:alignment="4">
   <element dfdl:alignment="4" .fixed length element of length 4.../>
   <element dfdl:alignment="8" .fixed length element of length 8 ../>
</sequence>
```
because the 2nd element may or may not require 4 bytes of alignment fill depending on the actual alignment of the starting sequence. This is a complete pain in the neck because you can't even tell the length of this structure without knowing where it starts despite the fact that it is all fixed length stuff.  (This is called the "interior aligment problem" lots of things that appear fixed length become variable length.)

I think we should have an option where we just say no to this. We have a property dfdlx:alignmentKind. Current values are 'manual' (no alignment regions ever inserted) or 'automatic' (everything including interior alignment allowed). 

I'm suggesting we add another enum value for this of 'onlyForVariableLength' which says alignment must be achieved manually by the schema author (using things like leading/trailing skip properties) except after variable-length elements.

Daffodil's core isKnownToBeAligned is false for the 2nd element in the example above even if it is true for the surrounding sequence. 

So when dfdl:alignmentKind is 'onlyForVariableLength' we would make it an SDE if any element is not knownToBeAligned unless it is preceded by something variable length. The SDE could require the author to increase the alignment of the surrounding sequence to 8, and then they could fix an SDE on the 2nd element by adding a leading skip  (or trailing to the prior).

Reply:
That's a worthwhile idea for a future PR, not part of this PR.